### PR TITLE
Master to PostgreSQL version 16

### PR DESCRIPTION
**Pull request by jrgemignani**
** **PLEASE DO NOT MERGE** ** until thoroughly reviewed and approved by all reviewers. Reviewers must state their approval as a comment in the thread below.

PR for moving the **master** branch up to PostgreSQL version 16.

It contains all of the latest work in the master branch in addition to all of the changes made to go from PostgreSQL version 15 to 16. When this PR is merged the master will be the latest and on PostgreSQL version 16.

Everyone listed as a reviewer needs to review this PR

### CODE REVIEW

File: src/backend/catalog/ag_label.c [hunk 1]
```diff
@@ -188,8 +188,9 @@ Datum _label_name(PG_FUNCTION_ARGS)
     uint32 label_id;
 
     if (PG_ARGISNULL(0) || PG_ARGISNULL(1))
-        ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
```
Comment (jrgemignani) [inline src/backend/catalog/ag_label.c:191]: This error message should not be changed. It hides an issue by having it return NULL.

File: src/backend/utils/adt/agtype.c [hunk 2]
```diff
@@ -2191,9 +2193,12 @@ Datum _agtype_build_vertex(PG_FUNCTION_ARGS)
     /* handles null */
     if (fcinfo->args[0].isnull)
     {
+        /*
         ereport(ERROR,
                 (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                  errmsg("_agtype_build_vertex() graphid cannot be NULL")));
+                 */
+        PG_RETURN_NULL();
```
Comment (jrgemignani) [inline src/backend/utils/adt/agtype.c:2201]: Why?

File: src/backend/utils/adt/agtype.c [hunk 3]
```diff
@@ -2262,9 +2267,12 @@ Datum _agtype_build_edge(PG_FUNCTION_ARGS)
     /* process graph id */
     if (fcinfo->args[0].isnull)
     {
+        PG_RETURN_NULL();
```
Comment (jrgemignani) [inline src/backend/utils/adt/agtype.c:2270]: Why?

File: src/backend/utils/adt/agtype.c [hunk 4]
```diff
@@ -2961,7 +2969,7 @@ Datum agtype_to_int2(PG_FUNCTION_ARGS)
 
     PG_FREE_IF_COPY(arg_agt, 0);
 
-    PG_RETURN_INT16(result);
+    PG_RETURN_INT64(result);
```
Comment (jrgemignani) [inline src/backend/utils/adt/agtype.c:2972]: Why? result is an int16

File: src/backend/utils/adt/agtype.c [hunk 5]
```diff
@@ -4268,7 +4278,7 @@ Datum agtype_hash_cmp(PG_FUNCTION_ARGS)
     uint64 seed = 0xF0F0F0F0;
 
     if (PG_ARGISNULL(0))
-        PG_RETURN_INT16(0);
+        PG_RETURN_INT64(0);
```
Comment (jrgemignani) [inline src/backend/utils/adt/agtype.c:4281]: Why? Function is expected to return integer (int16)

File: src/backend/executor/cypher_utils.c [hunk 6]
```diff
@@ -77,7 +77,7 @@ ResultRelInfo *create_entity_result_rel_info(EState *estate, char *graph_name,
 
     // initialize the resultRelInfo
     InitResultRelInfo(resultRelInfo, label_relation,
-                      list_length(estate->es_range_table), NULL,
+                      0, NULL,
```
Comment (jrgemignani) [inline src/backend/executor/cypher_utils.c:80]: Please explain

File: src/backend/utils/adt/agtype.c [hunk 7]
```diff
@@ -6014,7 +6024,7 @@ Datum age_tointeger(PG_FUNCTION_ARGS)
              */
             result = float8in_internal_null(string, NULL, "double precision",
                                             string, &is_valid);
-
+
```
Comment (rafsun42) [inline src/backend/utils/adt/agtype.c:6027]: Extra space

File: src/backend/utils/adt/agtype_parser.c [hunk 8]
```diff
@@ -30,15 +30,18 @@
...
 #include "libpq/pqformat.h"
 #include "miscadmin.h"
 #include "utils/date.h"
 #include "utils/datetime.h"
+#include "utils/varlena.h"
 
+#include "utils/agtype.h"
 #include "utils/agtype_parser.h"
 
+
```
Comment (rafsun42) [inline src/backend/utils/adt/agtype_parser.c:44]: Extra newline

**Review by rafsun42 — CHANGES_REQUESTED** [review]

**Review by MuhammadTahaNaveed — CHANGES_REQUESTED** [review]

File: src/backend/catalog/ag_catalog.c [hunk 9]
```diff
@@ -97,8 +97,15 @@ void ag_ProcessUtility_hook(PlannedStmt *pstmt, const char *queryString, bool re
...
     else
+    {
+        Assert(IsA(pstmt, PlannedStmt));
+        Assert(pstmt->commandType == CMD_UTILITY);
+        Assert(queryString != NULL);	/* required as of 8.4 */
+        Assert(qc == NULL || qc->commandTag == CMDTAG_UNKNOWN);
         standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params, queryEnv,
                                 dest, qc);
+    }
+
```
Comment (MuhammadTahaNaveed) [inline src/backend/catalog/ag_catalog.c:108]: extra line ?

**Review by rafsun42 — APPROVED** [review]

**Comment by jrgemignani** [comment]
@MuhammadTahaNaveed @muhammadshoaib @dehowef @rafsun42 Please review this PR again as there have been many additional changes.

**Review by rafsun42 — APPROVED** [review]
Reviewed it again. It looks ready to go!

File: .gitignore
Comment (dehowef) [inline .gitignore:1]: Is it okay to edit the gitignore?

**Review by dehowef — APPROVED** [review]

**Review by muhammadshoaib — APPROVED** [review]

File: src/backend/parser/cypher_clause.c [hunk 10]
```diff
@@ -7085,6 +7121,50 @@ static FuncExpr *make_clause_func_expr(char *function_name,
...
+/*
+ * This function is borrowed from PG version 16.1.
+ *
+ * It is used in transformations involving left join in Optional Match and
+ * Merge in a similar way PG16's transformFromClauseItem() uses it.
+ */
+static void markRelsAsNulledBy(ParseState *pstate, Node *n, int jindex)
+{
+    int			varno;
+    ListCell   *lc;
```
Comment (MuhammadTahaNaveed) [inline src/backend/parser/cypher_clause.c:7133]: extra spaces between datatype and variable name.

**Review by MuhammadTahaNaveed — APPROVED** [review]
Other than a minor cosmetic improvement pointed out, looks good to me 👍🏻 .

File: .gitignore
Comment (jrgemignani) [reply .gitignore:1]: We should probably look into the gitignore file and create a PR to clean it up as necessary.

**Comment by jirutka** [comment]
Great! Can you please release a new version now? :)

**Comment by jrgemignani** [comment]
@jirutka I know you're anxious :) but before we can release a new version we need to fold the changes to the **master** branch back into the **PG16** branch. At least, for **PG16** to be released. That likely won't happen until the end of next week, due to holidays. Then we can start preparing for the release.

**Comment by jrgemignani** [comment]
@jirutka That work starts with #1454

**Comment by jrgemignani** [comment]
@jirutka That process will be started next week as PG16 is now ready for release.

**Pull request merged**